File: voicetyping/ios/Shared/LLMFormatter.swift
import Foundation

final class LLMFormatter {

    private let settings: Settings
    private let regexCleanup: RegexCleanup
    private let session: URLSession

    init(settings: Settings = .shared, session: URLSession = .shared) {
        self.settings = settings
        self.regexCleanup = RegexCleanup()
        self.session = session
    }

    func format(_ text: String, mode: OutputMode) async -> FormattingResult {
        guard !text.isEmpty else {
            return FormattingResult(original: text, cleaned: text, mode: mode, wasLLMFormatted: false)
        }

        if mode == .raw {
            return FormattingResult(original: text, cleaned: text, mode: mode, wasLLMFormatted: false)
        }

        // Try LLM first
        do {
            let cleaned = try await callProxy(text: text, mode: mode)
            return FormattingResult(original: text, cleaned: cleaned, mode: mode, wasLLMFormatted: true)
        } catch {
            // Fallback to regex
            let cleaned = regexCleanup.clean(text)
            return FormattingResult(original: text, cleaned: cleaned, mode: mode, wasLLMFormatted: false)
        }
    }

    private func callProxy(text: String, mode: OutputMode) async throws -> String {
        guard let url = URL(string: settings.proxyURL) else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 5

        let body: [String: String] = [
            "text": text,
            "mode": mode.rawValue,
            "deviceId": settings.deviceId,
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let result = json["result"] as? String else {
            throw URLError(.cannotParseResponse)
        }

        return result
    }
}

File: voicetyping/android/app/src/main/kotlin/com/voicetyping/formatter/LLMFormatter.kt
package com.voicetyping.formatter

import com.voicetyping.model.OutputMode
import com.voicetyping.model.Settings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

data class FormattingResult(
    val original: String,
    val cleaned: String,
    val mode: OutputMode,
    val wasLLMFormatted: Boolean,
)

class LLMFormatter(private val settings: Settings) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(5, TimeUnit.SECONDS)
        .build()

    private val regexCleanup = RegexCleanup()

    suspend fun format(text: String, mode: OutputMode): FormattingResult {
        if (text.isBlank()) return FormattingResult(text, text, mode, false)
        if (mode == OutputMode.RAW) return FormattingResult(text, text, mode, false)

        return try {
            val cleaned = callProxy(text, mode)
            FormattingResult(text, cleaned, mode, true)
        } catch (_: Exception) {
            val cleaned = regexCleanup.clean(text)
            FormattingResult(text, cleaned, mode, false)
        }
    }

    private suspend fun callProxy(text: String, mode: OutputMode): String =
        withContext(Dispatchers.IO) {
            val json = JSONObject().apply {
                put("text", text)
                put("mode", mode.name.lowercase())
                put("deviceId", settings.deviceId)
            }

            val body = json.toString().toRequestBody("application/json".toMediaType())
            val request = Request.Builder()
                .url(settings.proxyURL)
                .post(body)
                .build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) throw RuntimeException("HTTP ${response.code}")

            val responseBody = response.body?.string() ?: throw RuntimeException("Empty response")
            JSONObject(responseBody).getString("result")
        }
}
